staging/src/k8s.io/client-go/tools/events/event_broadcaster.go

```diff
@@ -0,0 +1,302 @@
+	"k8s.io/apimachinery/pkg/util/clock"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	restclient "k8s.io/client-go/rest"
+
+	"os"
+
+	"k8s.io/api/events/v1beta1"
+	"k8s.io/apimachinery/pkg/util/json"
+	"k8s.io/apimachinery/pkg/util/strategicpatch"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/record/util"
+	"k8s.io/klog"
+)
+
+const (
+	maxTriesPerEvent = 12
+	finishTime       = 6 * time.Minute
+	refreshTime      = 30 * time.Minute
+)
+
+const maxQueuedEvents = 1000
+
+var defaultSleepDuration = 10 * time.Second
+
+type eventKey struct {
+	action              string
+	reason              string
+	reportingController string
+	reportingInstance   string
+	regarding           corev1.ObjectReference
+	related             corev1.ObjectReference
+}
+
+type eventBroadcasterImpl struct {
+	*watch.Broadcaster
+	eventCache    map[eventKey]*v1beta1.Event
+	mu            sync.RWMutex
+	sleepDuration time.Duration
+	sink          EventSink
+}
+
+// NewBroadcaster Creates a new event broadcaster.
+func NewBroadcaster(sink EventSink) EventBroadcaster {
+	return newBroadcaster(sink, defaultSleepDuration)
+}
+
+// NewBroadcasterForTest Creates a new event broadcaster for test purposes.
+func newBroadcaster(sink EventSink, sleepDuration time.Duration) EventBroadcaster {
+	return &eventBroadcasterImpl{
+		Broadcaster:   watch.NewBroadcaster(maxQueuedEvents, watch.DropIfChannelFull),
+		eventCache:    map[eventKey]*v1beta1.Event{},
+		sleepDuration: sleepDuration,
+		sink:          sink,
+	}
+}
+
+func (e *eventBroadcasterImpl) refreshExistingEventSeries() {
+	e.mu.RLock()
+	defer e.mu.RUnlock()
+	for _, event := range e.eventCache {
+
+		if event.Series != nil {
+			//TODO: investigate locking
+			if recordedEvent, retry := recordEvent(e.sink, event); !retry {
+				event = recordedEvent
+			}
+		}
+
+	}
+}
+
+func (e *eventBroadcasterImpl) finishSeries() {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	for isomorphicKey, event := range e.eventCache {
+		eventSerie := event.Series
+		if eventSerie != nil {
+			if eventSerie.LastObservedTime.Time.Add(finishTime).Before(time.Now()) {
+				//TODO: investigate locking
+				if recordedEvent, retry := recordEvent(e.sink, event); !retry {
+					event = recordedEvent
+					delete(e.eventCache, isomorphicKey)
+				}
+			}
+		}
+
+	}
+}
+
+// NewRecorder returns an EventRecorder that records events with the given event source.
+func (e *eventBroadcasterImpl) NewRecorder(scheme *runtime.Scheme, reportingController string) EventRecorder {
+
+	hostname, _ := os.Hostname()
+	reportingInstance := reportingController + "-" + hostname
+
+	return &recorderImpl{scheme, reportingController, reportingInstance, e.Broadcaster, clock.RealClock{}}
+}
+
+func (e *eventBroadcasterImpl) recordToSink(event *v1beta1.Event, clock clock.Clock) {
+	// Make a copy before modification, because there could be multiple listeners.
+	// Events are safe to copy like this.
```

Comment:
They aren't, e.g.:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/events/v1beta1/types.go#L64

Reply:
@wojtek-t  - Seems like we should deepcopy `Related`, As we cannot assume that no one is mutating the isomorphic key fields, thoughts ?

Reply:
We should DeepCopy the whole object. Just do:
```
eventCopy := event.DeepCopyObject()
```
[can't remember for sure, you may also need some casting]